File: frontend/src/utils/inferenceHowToMint.ts
/** Auto-mint an inference key when the Ollama howto card opens (GitHub #225). */

export const INFERENCE_HOWTO_AUTO_NAME = 'Ollama howto (auto)'
export const INFERENCE_HOWTO_KIND = 'inference'
export const INFERENCE_HOWTO_EXPIRES_IN = '90d'
export const INFERENCE_HOWTO_SIGN_IN = 'Sign in required'

export type InferenceHowToKeyRow = {
  kind?: string | null
}

export function hasInferenceKey(keys: InferenceHowToKeyRow[]): boolean {
  return keys.some((key) => key.kind === INFERENCE_HOWTO_KIND)
}

export function needsInferenceHowToMint(keys: InferenceHowToKeyRow[]): boolean {
  return !hasInferenceKey(keys)
}

export function inferenceHowToMintBody(): {
  name: string
  expiresIn: string
  kind: string
} {
  return {
    name: INFERENCE_HOWTO_AUTO_NAME,
    expiresIn: INFERENCE_HOWTO_EXPIRES_IN,
    kind: INFERENCE_HOWTO_KIND,
  }
}

/** Session/auth failure copy. 403 is not treated as "not admin". */
export function inferenceHowToMintBanner(input: {
  status?: number | null
  message?: string | null
}): string {
  if (input.status === 401) return INFERENCE_HOWTO_SIGN_IN
  const message = (input.message ?? '').trim()
  return message || 'Could not mint an inference key'
}

File: Apps/BarkVisorConsole/Sources/Models/InferenceHowToMint.swift
import Foundation

/// Auto-mint an inference key when the Ollama howto card opens (GitHub #225).
enum InferenceHowToMint {
    static let autoName = "Ollama howto (auto)"
    static let kind = APIKeyKindOption.inference.rawValue
    static let expiresIn = APIKeyDisplay.defaultExpiry

    static func needsMint(keys: [APIKeyResponse], now: Date = Date()) -> Bool {
        !keys.contains { $0.kind == kind && !isExpired($0.expiresAt, now: now) }
    }

    static func isExpired(_ expiresAt: String?, now: Date = Date()) -> Bool {
        guard let expiresAt else { return false }
        guard let date = APIKeyDisplay.parseISO8601(expiresAt) else { return false }
        return date < now
    }

    static func createBody() -> APIKeyCreateBody {
        APIKeyCreateBody(name: autoName, expiresIn: expiresIn, kind: kind)
    }

    /// Session/auth failure copy. 403 is not treated as "not admin".
    static func bannerMessage(from error: Error) -> String {
        guard let api = error as? APIError else {
            return error.localizedDescription
        }
        switch api {
        case .unauthorized:
            return APIKeyDisplay.signInRequired
        case let .http(status, _) where status == 401:
            return APIKeyDisplay.signInRequired
        default:
            return api.localizedDescription
        }
    }
}
